File: gems/decomplex/rust/src/decomplex/ast/adapters/lua.rs
use super::super::{
    bracketed, direct_binary_operator, lua_keyed_table_target, lua_positional_table_target,
    named_children, node_text, raw_named_children, LUA_LEADING_FUNCTION_WRAPPER_KINDS,
    LUA_LEADING_IF_WRAPPER_KINDS,
};
use super::base::{AstNormalizationAdapter, NamedChildrenAction, LUA_ASSIGNMENT_OPERATORS};
use tree_sitter::Node as TreeSitterNode;

pub(crate) struct LuaAstAdapter;

impl AstNormalizationAdapter for LuaAstAdapter {
    fn explicit_alternative<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
    ) -> Option<TreeSitterNode<'tree>> {
        named_children(node)
            .into_iter()
            .find(|child| matches!(child.kind(), "elseif_statement" | "else" | "else_statement"))
    }

    fn ternary_parts<'tree>(
        &self,
        _node: TreeSitterNode<'tree>,
        _source: &str,
    ) -> Option<super::super::TernaryParts<'tree>> {
        None
    }

    fn named_children_action<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
        children: &[TreeSitterNode<'tree>],
    ) -> NamedChildrenAction<'tree> {
        if node.kind() == "variable_list" && children.len() == 1 {
            if children[0].kind() == "identifier" && lua_single_assignment_block_child(node, source)
            {
                return NamedChildrenAction::Drop;
            }
            if node
                .parent()
                .map(|parent| parent.kind() == "for_generic_clause")
                .unwrap_or(false)
            {
                return NamedChildrenAction::Drop;
            }
            if node
                .parent()
                .map(|parent| {
                    parent.kind() == "variable_declaration" && raw_named_children(parent).len() == 1
                })
                .unwrap_or(false)
            {
                return NamedChildrenAction::Drop;
            }
            if children[0].kind() == "dot_index_expression"
                && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Recurse(children[0]);
            }
        }

        if node.kind() == "expression_list" && children.len() == 1 {
            if children[0].kind() == "identifier"
                && node
                    .parent()
                    .map(|parent| {
                        matches!(parent.kind(), "assignment_statement" | "return_statement")
                    })
                    .unwrap_or(false)
                && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Drop;
            }
            if matches!(
                children[0].kind(),
                "true" | "false" | "nil" | "number" | "integer" | "float"
            ) && node
                .parent()
                .map(|parent| matches!(parent.kind(), "assignment_statement" | "return_statement"))
                .unwrap_or(false)
                && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Drop;
            }
            if matches!(
                children[0].kind(),
                "binary_expression"
                    | "function_call"
                    | "dot_index_expression"
                    | "function_definition"
                    | "string"
                    | "table_constructor"
            ) && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Recurse(children[0]);
            }
        }

        if node.kind() == "field" && children.len() == 1 {
            if children[0].kind() == "identifier"
                && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Drop;
            }
            if matches!(children[0].kind(), "string" | "function_call")
                && node_text(node, source) == node_text(children[0], source)
            {
                return NamedChildrenAction::Recurse(children[0]);
            }
        }

        if node.kind() == "block"
            && children.len() == 1
            && matches!(
                children[0].kind(),
                "function_call" | "return_statement" | "variable_declaration"
            )
            && node_text(node, source) == node_text(children[0], source)
        {
            return NamedChildrenAction::Recurse(children[0]);
        }

        NamedChildrenAction::Default
    }

    fn unary_minus_expression(&self, node: TreeSitterNode<'_>, source: &str) -> bool {
        (matches!(node.kind(), "unary" | "unary_expression")
            && node_text(node, source).trim_start().starts_with('-'))
            || (node.kind() == "expression_list"
                && node
                    .children(&mut node.walk())
                    .next()
                    .map(|child| node_text(child, source) == "-")
                    .unwrap_or(false)
                && named_children(node).len() == 1)
    }

    fn binary_operator(&self, node: TreeSitterNode<'_>, source: &str) -> Option<String> {
        if let Some(operator) = direct_binary_operator(node, source) {
            return Some(operator.to_string());
        }

        let child = self.exact_single_named_child(node, self.binary_wrapper_kinds(), source)?;
        self.binary_operator(child, source)
    }

    fn unwrap_node(
        &self,
        node: TreeSitterNode<'_>,
        source: &str,
        named_child_count: usize,
    ) -> bool {
        if matches!(
            node.kind(),
            "parenthesized_expression"
                | "parenthesized_statements"
                | "expression_statement"
                | "statement"
                | "case_pattern"
                | "match_pattern"
                | "pattern"
        ) && named_child_count == 1
        {
            return true;
        }

        if node.kind() != "expression_list" || named_child_count != 1 {
            return false;
        }

        let raw_named = raw_named_children(node);
        if raw_named.len() == 1
            && raw_named[0].kind() == "parenthesized_expression"
            && node_text(raw_named[0], source) == node_text(node, source)
        {
            return true;
        }

        let raw_children = node.children(&mut node.walk()).collect::<Vec<_>>();
        raw_children
            .first()
            .map(|child| node_text(*child, source) == "(")
            .unwrap_or(false)
            && raw_children
                .last()
                .map(|child| node_text(*child, source) == ")")
                .unwrap_or(false)
    }

    fn leading_function_statement(&self, node: TreeSitterNode<'_>, source: &str) -> bool {
        self.leading_function_statement_with_keyword(
            node,
            source,
            "function",
            LUA_LEADING_FUNCTION_WRAPPER_KINDS,
        )
    }

    fn leading_function_target<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if !LUA_LEADING_FUNCTION_WRAPPER_KINDS.contains(&node.kind()) {
            return None;
        }
        if node
            .children(&mut node.walk())
            .next()
            .map(|child| child.kind() == "function")
            .unwrap_or(false)
        {
            return Some(node);
        }
        self.exact_single_named_child(node, &["function_declaration"], source)
    }

    fn leading_function_body_kind(&self) -> &'static str {
        "block"
    }

    fn leading_if_target<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if LUA_LEADING_IF_WRAPPER_KINDS.contains(&node.kind()) {
            if let Some(child) = self.exact_single_named_child(node, &["if_statement"], source) {
                return Some(child);
            }
        }
        if super::super::LEADING_IF_WRAPPER_KINDS.contains(&node.kind()) {
            return Some(node);
        }
        None
    }

    fn array_literal_target<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if let Some(target) = lua_positional_table_target(node, source) {
            return Some(target);
        }

        if super::super::ARRAY_LITERAL_NODE_KINDS.contains(&node.kind()) {
            return Some(node);
        }
        if !super::super::ARRAY_LITERAL_WRAPPER_KINDS.contains(&node.kind()) {
            return None;
        }
        if bracketed(node, source, "[", "]") {
            return Some(node);
        }
        let named = named_children(node);
        let child = *named.first()?;
        if named.len() == 1 {
            if super::super::ARRAY_LITERAL_NODE_KINDS.contains(&child.kind()) {
                return Some(child);
            }
            if matches!(child.kind(), "expression_statement" | "statement")
                && node_text(child, source).trim() == node_text(node, source).trim()
            {
                return self.array_literal_target(child, source);
            }
        }
        None
    }

    fn array_literal_values<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Vec<TreeSitterNode<'tree>> {
        let target = self.array_literal_target(node, source).unwrap_or(node);
        if target.kind() == "arguments" {
            if let Some(table) = named_children(target)
                .into_iter()
                .find(|child| child.kind() == "table_constructor")
            {
                if node_text(target, source).trim() == node_text(table, source).trim() {
                    return named_children(table);
                }
            }
        }
        if target.kind() == "table_constructor" {
            return named_children(target);
        }

        named_children(target)
    }

    fn hash_literal_target<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if let Some(target) = lua_keyed_table_target(node, source) {
            return Some(target);
        }

        if super::super::HASH_LITERAL_NODE_KINDS.contains(&node.kind()) {
            return Some(node);
        }
        if !super::super::HASH_LITERAL_WRAPPER_KINDS.contains(&node.kind()) {
            return None;
        }
        None
    }

    fn hash_literal_values<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Vec<TreeSitterNode<'tree>> {
        let target = self.hash_literal_target(node, source).unwrap_or(node);
        if target.kind() == "arguments" {
            if let Some(table) = named_children(target)
                .into_iter()
                .find(|child| child.kind() == "table_constructor")
            {
                return named_children(table);
            }
            return named_children(target);
        }
        if target.kind() == "table_constructor" {
            return named_children(target);
        }

        named_children(target)
    }

    fn member_assignment_target(&self, node: TreeSitterNode<'_>, source: &str) -> bool {
        if node.kind() != "variable_list" {
            return false;
        }

        let raw_named = raw_named_children(node);
        let target = if raw_named.len() == 1
            && raw_named[0].kind() == "dot_index_expression"
            && node_text(node, source) == node_text(raw_named[0], source)
        {
            raw_named[0]
        } else {
            node
        };

        raw_named_children(target).len() == 2
            && target
                .children(&mut target.walk())
                .any(|child| !child.is_named() && node_text(child, source) == ".")
    }

    fn literal_fragment_assignment_context(&self, node: TreeSitterNode<'_>, source: &str) -> bool {
        let Some(parent) = node.parent() else {
            return false;
        };
        if matches!(
            parent.kind(),
            "string" | "delimited_symbol" | "regex" | "regex_literal"
        ) {
            return true;
        }
        matches!(
            node.kind(),
            "string_content" | "escape_sequence" | "interpolation" | "string_fragment"
        ) && (parent.kind() == "expression_list"
            || parent
                .parent()
                .map(|grandparent| {
                    matches!(
                        grandparent.kind(),
                        "string" | "delimited_symbol" | "regex" | "regex_literal"
                    )
                })
                .unwrap_or(false))
            && !source.is_empty()
    }

    fn lambda_target<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if node.kind() == "function_definition" {
            return Some(node);
        }

        if node.kind() == "expression_list" {
            if node
                .children(&mut node.walk())
                .next()
                .map(|child| child.kind() == "function")
                .unwrap_or(false)
                && named_children(node)
                    .iter()
                    .any(|child| child.kind() == "block")
            {
                return Some(node);
            }

            let named = named_children(node);
            if named.len() == 1
                && named[0].kind() == "function_definition"
                && node_text(named[0], source) == node_text(node, source)
            {
                return Some(named[0]);
            }
        }

        if node.kind() == "lambda" {
            Some(node)
        } else {
            None
        }
    }

    fn operator_call_expression_kind(&self, node: TreeSitterNode<'_>) -> bool {
        matches!(
            node.kind(),
            "binary" | "binary_expression" | "expression_list"
        )
    }

    fn boolean_expression_kind(&self, node: TreeSitterNode<'_>) -> bool {
        super::super::BOOLEAN_EXPRESSION_KINDS.contains(&node.kind())
            || node.kind() == "expression_list"
    }

    fn comparison_expression_kind(&self, node: TreeSitterNode<'_>) -> bool {
        super::super::COMPARISON_EXPRESSION_KINDS.contains(&node.kind())
            || node.kind() == "expression_list"
    }

    fn single_assignment_block_child(&self, node: TreeSitterNode<'_>, source: &str) -> bool {
        lua_single_assignment_block_child(node, source)
    }

    fn member_read_excluded(&self, node: TreeSitterNode<'_>) -> bool {
        node.kind() == "field"
    }

    fn no_paren_string_argument_content<'tree>(
        &self,
        node: TreeSitterNode<'tree>,
        source: &str,
    ) -> Option<TreeSitterNode<'tree>> {
        if node.kind() != "string" {
            return None;
        }
        let parent = node.parent()?;
        if parent.kind() != "arguments" || node_text(parent, source) != node_text(node, source) {
            return None;
        }
        raw_named_children(node)
            .into_iter()
            .find(|child| child.kind() == "string_content")
    }

    fn assignment_operators(&self) -> &'static [&'static str] {
        LUA_ASSIGNMENT_OPERATORS
    }
}

fn lua_single_assignment_block_child(node: TreeSitterNode<'_>, source: &str) -> bool {
    let Some(parent) = node.parent() else {
        return false;
    };
    if parent.kind() != "assignment_statement" {
        return false;
    }
    let Some(grandparent) = parent.parent() else {
        return false;
    };
    grandparent.kind() == "block"
        && node_text(grandparent, source) == node_text(parent, source)
        && raw_named_children(grandparent).len() == 1
}

File: gems/fact-mine/lib/fact_mine/ast/adapters/lua.rb
# frozen_string_literal: true

require_relative "base"

module FactMine
  module Ast
    class LuaTreeSitterNormalizationAdapter < TreeSitterNormalizationAdapter
      ASSIGNMENT_OPERATORS = %w[=].freeze
      LEADING_FUNCTION_WRAPPER_KINDS = %w[block].freeze
      LEADING_IF_WRAPPER_KINDS = %w[block].freeze

      def explicit_alternative(node)
        node.named_children.find { |child| %w[elseif_statement else else_statement].include?(child.kind) }
      rescue StandardError
        nil
      end

      def unary_minus_expression?(node)
        super ||
          (node.kind == "expression_list" && node.children.first&.text == "-" && node.named_children.size == 1)
      rescue StandardError
        false
      end

      def binary_operator(node)
        direct = direct_binary_operator(node)
        return direct.to_s if direct

        child = exact_single_named_child(node, kinds: BINARY_WRAPPER_KINDS)
        child ? binary_operator(child) : ""
      end

      def unwrap_node?(node)
        super ||
          (node.kind == "expression_list" &&
            node.named_children.size == 1 &&
            node.children.first&.text == "(" &&
            node.children.last&.text == ")")
      rescue StandardError
        false
      end

      def leading_function_statement?(node)
        leading_function_statement_with_keyword?(node, "function", LEADING_FUNCTION_WRAPPER_KINDS)
      end

      def leading_function_body(node)
        node.named_children.reverse.find { |child| child.kind == "block" }
      rescue StandardError
        nil
      end

      def leading_if_target(node)
        if LEADING_IF_WRAPPER_KINDS.include?(node.kind)
          child = exact_single_named_child(node, kinds: %w[if_statement])
          return child if child
        end

        super
      end

      def special_if_statement?(node)
        node.kind == "if_statement"
      rescue StandardError
        false
      end

      def normalize_special_if(node, helpers:)
        named = node.named_children
        cond = named.first
        positive = named.find { |child| child.kind == "block" }
        alternatives = named.drop_while { |child| child != positive }.drop(1)
        helpers.__send__(
          :wrap,
          :IF,
          children: [
            helpers.__send__(:normalize_node, cond),
            helpers.__send__(:normalize_body, positive),
            normalize_lua_alternatives(alternatives, helpers: helpers)
          ],
          source: node
        )
      end

      def array_literal_target(node)
        if node.kind == "block"
          named = node.named_children
          if named.size == 2 && named.first.kind == "identifier" && named.first.text.to_s.empty?
            target = lua_positional_table_arguments(named[1])
            return target if target
          end
        end

        target = lua_positional_table_arguments(node)
        return target if target

        super
      rescue StandardError
        nil
      end

      def hash_literal_target(node)
        target = lua_keyed_table_arguments(node)
        return target if target

        super
      rescue StandardError
        nil
      end

      def hash_literal_values(node)
        target = hash_literal_target(node) || node
        return target.named_children if target.kind == "arguments"

        super
      rescue StandardError
        []
      end

      def identifier_text_node?(node)
        %w[variable_list expression_list].include?(node.kind) &&
          node.text.to_s.match?(/\A[A-Za-z_]\w*\z/)
      rescue StandardError
        false
      end

      def member_assignment_target?(node)
        return false unless node.kind == "variable_list"

        node.named_children.size == 2 &&
          node.children.any? { |child| !child.named? && child.text == "." }
      rescue StandardError
        false
      end

      def literal_fragment_assignment_context?(node)
        return true if super

        literal_fragment_kind?(node) && node.parent&.kind == "expression_list"
      rescue StandardError
        false
      end

      def lambda_target(node)
        return node if node.kind == "function_definition"

        if node.kind == "expression_list"
          return node if node.children.first&.kind == "function" &&
            node.named_children.any? { |child| child.kind == "block" }

          named = node.named_children
          return named.first if named.size == 1 && named.first.kind == "function_definition"
        end

        super
      rescue StandardError
        nil
      end

      private

      def normalize_lua_alternatives(nodes, helpers:)
        return nil if nodes.empty?

        current = nodes.first
        rest = nodes.drop(1)
        if current.kind == "elseif_statement"
          cond = current.named_children.first
          positive = current.named_children.find { |child| child.kind == "block" }
          return helpers.__send__(
            :wrap,
            :IF,
            children: [
              helpers.__send__(:normalize_node, cond),
              helpers.__send__(:normalize_body, positive),
              normalize_lua_alternatives(rest, helpers: helpers)
            ],
            source: current
          )
        end
        return helpers.__send__(:normalize_else_or_branch, current) if current.kind == "else_statement"

        normalize_lua_alternatives(rest, helpers: helpers)
      end

      def lua_positional_table_arguments(node)
        return nil unless node&.kind == "arguments"
        return nil unless bracketed?(node, "{", "}")

        fields = node.named_children
        return nil if fields.empty?
        return nil unless fields.all? { |field| field.kind == "field" && field.named_children.size <= 1 }

        node
      end

      def lua_keyed_table_arguments(node)
        if node&.kind == "block"
          named = node.named_children
          if named.size == 2 && named.first.kind == "identifier" && named.first.text.to_s.empty?
            return lua_keyed_table_arguments(named[1])
          end
        end

        return nil unless node&.kind == "arguments"
        return nil unless bracketed?(node, "{", "}")

        fields = node.named_children
        return node if fields.empty?
        return nil if fields.all? { |field| field.kind == "field" && field.named_children.size <= 1 }

        node
      end

      private

      def assignment_operators
        ASSIGNMENT_OPERATORS
      end

      def operator_call_expression_kinds
        super + %w[expression_list]
      end

      def boolean_expression_kinds
        super + %w[expression_list]
      end

      def comparison_expression_kinds
        super + %w[expression_list]
      end
    end

  end
end
